connector/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala

```diff
@@ -316,6 +320,50 @@ private[kafka010] class KafkaMicroBatchStream(
     }
   }
 
+  private def assertEndOffsetForTriggerAvailableNow(
+      endPartitionOffsets: Map[TopicPartition, Long]): Unit = {
+    val tpsForPrefetched = allDataForTriggerAvailableNow.keySet
+    val tpsForEndOffset = endPartitionOffsets.keySet
+
+    if (tpsForPrefetched != tpsForEndOffset) {
+      throw KafkaExceptions.topicPartitionsInEndOffsetAreNotSameWithPrefetched(
+        tpsForPrefetched, tpsForEndOffset)
+    }
+
+    val endOffsetHasGreaterThanPrefetched = {
+      allDataForTriggerAvailableNow.keySet.exists { tp =>
+        val offsetFromPrefetched = allDataForTriggerAvailableNow(tp)
+        val offsetFromEndOffset = endPartitionOffsets(tp)
+        offsetFromEndOffset > offsetFromPrefetched
+      }
+    }
+    if (endOffsetHasGreaterThanPrefetched) {
+      throw KafkaExceptions.endOffsetHasGreaterOffsetForTopicPartitionThanPrefetched(
+        allDataForTriggerAvailableNow, endPartitionOffsets)
+    }
+
+    val latestOffsets = kafkaOffsetReader.fetchLatestOffsets(Some(endPartitionOffsets))
```

Comment:
Why do we need to fetch the latest offset from Kafka again?  This will add additional latency.  I know its trigger available now but what does this buy us?

Reply:
If we don't fetch the latest offset with latest topic-partitions from Kafka again, what we are trying to guard against? If someone turns on failOnDataLoss and runs the query with Trigger.AvailableNow, it will just move on when specific topic partition is dropped, leading that it never reaches the end state (prepared offset).

Reply:
It's not safe to just pretend that Kafka resource won't change during the query run. If it happens, it will end up with unexpected behavior e.g. consumer based grouping, polling from non-exist topic partition may lead to timeout of metadata fetch.

Reply:
These things can happen regardless of whether availableNow trigger is used right?  What happens if these things happen when using processing time trigger?

Reply:
If failOnDataLoss is turned on, we tolerate the removal of topic partition, and allow offset to go "backward".

Reply:
The thing is, when the query is running with AvailableNow, there is an "end" condition, and if the offset is off against the direction to reach an end condition, the query will run infinitely. Huge difference with other triggers. And also, in other triggers, they fetch the latest information in each batch already.

Reply:
> If failOnDataLoss is turned on, we tolerate the removal of topic partition, and allow offset to go "backward".

You mean 'if not turned on'? If we tolerate with 'processing time' trigger, any reason why we shouldn't tolerate here (and avoid waiting forever). 
E.g. we could reset the offsets.
I am just thinking aloud, we don't need to block on this.

Reply:
Yeah sorry I meant if it's not turned on (turned off).

> If we tolerate with 'processing time' trigger, any reason why we shouldn't tolerate here (and avoid waiting forever).
> E.g. we could reset the offsets.

Source may be able to do the smart thing, e.g. end offset being built by preparation can change based on the change of latest offset, but it would be also tricky. Maybe someone has bright idea to rebuild the end offset with consideration of failOnDataLoss, this would be better.